# RERConverge

RERConverge is a set of software written in R that estimates the correlation between relative evolutionary rates of genes and the evolution of a convergent binary or continuous trait across a phylogeny.

## Getting Started

Please refer to the [wiki](https://github.com/nclark-lab/RERconverge/wiki/Getting-Started) for instructions to **install** and **run** RERConverge on your own data.
<a href="https://github.com/nclark-lab/RERconverge/wiki/Getting-Started" target="_blank">New Tab</a>

## Authors

* **Maria Chikina** - [mchikina](https://github.com/mchikina)
* **Nathan Clark** - [nclark-lab](https://github.com/nclark-lab)
* **Amanda Kowalczyk** - [Amylith](https://github.com/Amylith)
* **Weiguang (Wayne) Mao** - [wgmao](https://github.com/wgmao)
* **Wynn Meyer** - [sorrywm](https://github.com/sorrywm)
* **Raghavendran Partha** - [raghavendranpartha](https://github.com/raghavendranpartha)

See also the list of [contributors](https://github.com/raghavendranpartha/RERconverge/contributors) who participated in this project.

## Citation

RERConverge can be cited as follows:

```
For coding sequences:
Chikina M, Robinson JD, Clark NL. Hundreds of Genes Experienced Convergent Shifts in Selective Pressure in Marine 
Mammals. Mol Biol Evol. 2016;33: 2182–92. doi:10.1093/molbev/msw112

For conserved non-coding sequences:
Partha R, Chauhan B, Ferreira Z, Robinson J, Lathrop K, Nischal K, et al. Subterranean mammals show convergent 
regression in ocular genes and enhancers, along with adaptation to tunneling. In press eLife. 

```

## License

This project is licensed under the GNU General Public License v3.0 - see the [LICENSE](LICENSE) file for details

## Acknowledgments

* Methods for computing weights rely on ideas from the following paper:
```
Law CW, Chen Y, Shi W, Smyth GK. voom: Precision weights unlock linear model analysis tools for RNA-seq read
counts. Genome Biol. 2014;15: R29. doi:10.1186/gb-2014-15-2-r29

```
* Projection operations are drawn from the following paper:
```
Sato T, Yamanishi Y, Kanehisa M, Toh H. The inference of protein-protein interactions by co-evolutionary 
analysis is improved by excluding the information about the phylogenetic relationships. Bioinformatics. 
Bioinformatics Center, Institute for Chemical Research, Kyoto University, Gokasho, Uji, Kyoto 611-0011, 
Japan. sato@kuicr.kyoto-u.ac.jp; 2005;21: 3482–3489. doi:10.1093/bioinformatics/bti564

```
* Thanks to [PurpleBooth](https://github.com/PurpleBooth) for this template.
