---
title: "PermulationWalkthrough"
author: "Amanda Kowalczyk and Elysia Saputra"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This walkthrough provides instructions to perform permulation analysis to calculate empirical p-values for genes and pathways.  Due to a non-uniform p-value distribution for gene-evolutionary rate associations and non-independence among genes for pathway enrichment, parametric p-values directly from RERconverge may not accurately represent true confidence of association or enrichment.  Permulation p-values correct this issue.  For the continuous phenotype, parametric p-values for gene correlations tend to be overestimated and parametric p-values for pathway enrichment tend to be underestimated.

Permulations are performed using the following steps after performing standard RERconverge analyses:

1. generate null (permulated) phenotypes
2. recalculate gene correlation and pathway enrichment statistics using null phenotypes
3. quantify the proportion of null statistics more significant than observed statistics

Since statistics generated from null phenotypes represent the true null distribution for the statistic calculated over the given data, the proportion of extreme null statistics quantifies true confidence in the extremity of a particular correlation or enrichment.  These are permulation p-values and can be interpreted similarly to standard p-values, including performing multiple hypothesis testing correction.  One caveat is that the precision of a permulation p-value is limited by the number of permulations performed - the lowest observable p-value is the reciprocal of the number of permulations performed.

Permulations are a combination of phylogenetic simulations and permutations.  To generate a permulated phenotype, first a simulated phenotype is generated based on a phylogeny with branch lengths that represent the average genome-wide evolutionary rate along that branch using a Brownian motion model of evolution.  Observed phenotype values are then assigned to species based on the ranks of simulated values - the highest simulated value is assigned the highest observed value, the second-highest simulated values is assigned the second-highest observed value, etc.  Permulations are favored over permulations because they respect the underlying phylogenetic relationships among species, so more closely related species have more similar phenotypes.  Permulations are favored over simulations because they preserve the exact distribution and range of the observed phenotype.  In practice, permulation p-values are more conservative (i.e. lower) than permutation p-values and equally as conservative as simulation p-values.

**In R, permulation procedures are conveniently bundled into a handful of functions for simplicity.  Note that these functions can take a very long time to run on large datasets and for large numbers of permulations.**


# Continuous Permulations

First, conduct standard RERconverge analysis.  Please see full walkthroughs for more details about these steps.

```{r, message=F, warning=F, error=F, results=F, fig.show="hide"}
#load RER package
library(RERconverge)
rerpath = find.package('RERconverge')

#read trees
toytreefile = "subsetMammalGeneTrees.txt" 
toyTrees=RERconverge::readTrees(paste(rerpath,"/extdata/",toytreefile,sep=""), max.read = 200)

#load phenotype data
data("logAdultWeightcm") 

#calculate RERs
mamRERw = RERconverge::getAllResiduals(toyTrees,useSpecies=names(logAdultWeightcm), 
                                       transform = "sqrt", weighted = T, scale = T)

#generate trait tree
charpaths=RERconverge::char2Paths(logAdultWeightcm, toyTrees)

#calculate correlation statistics
res=RERconverge::correlateWithContinuousPhenotype(mamRERw, charpaths, min.sp = 10, 
                                                  winsorizeRER = 3, winsorizetrait = 3)

#calculate pathway enrichment statistics
stats=RERconverge::getStat(res)
annots=RERconverge::read.gmt("gmtfile.gmt")
annotlist=list(annots)
names(annotlist)="MSigDBpathways"
enrichment=RERconverge::fastwilcoxGMTall(stats, annotlist, outputGeneVals=T, num.g=10)

```

After calculating parametric statistics above, calculate permulations.  The `getPermsContinuous` function operates by generating null p-values and statistics for gene correlations and enrichment statistics.  The function requires the following input:

* `numperms`: the number of permulations to perform, recommended at least 1000.  Note that the total number of permulations is the limit to permulation p-value precion - the lowest possible permulation p-value is 1/numperms
* `traitvec`: phenotype vector as specified in `correlateWithContinuousPhenotype`
* `RERmat`: RER matrix from `getAllResiduals` as used in `correlateWithContinuousPhenotype`
* `annotlist`: annotations as used in `fastwilcoxGMTall`. Not used if `calculateenrich`=F
* `trees`: trees object from `readTrees` as used in `getAllResiduals`
* `mastertree`: rooted and fully dichotomous tree containing all species with branch lengths representing average evolutionary rate genome wide.  In most cases, modify the master tree from `trees`
* `calculateenrich`: default T.  Boolean specifying if permulation enrichment statistics should be calculated
* `type`: default "simperm". Specifies method to generate null phenotypes.  "simperm" specifies permulations, "sim" specifies phylogenetic simulations, and "perm" specifies permutations
* `winR` and `winT`: default 3.  Numeric values specifying how much to winsorize RER and trait trees, respectively.  Should match `winR` and `winT` values used in `correlateWithContinuousPhenotype`

**This example uses only 100 permulations as a toy example.  In practice, we have found that for this group of species and this gene set, at least 500 permulations should be performed to obtain meaningful p-values.  Ideally, as many permulations as possible should be performed to maximize p-value precision.**

```{r, message=F, warning=F, error=F, results=F}
mt=toyTrees$masterTree
mt=root.phylo(mt, outgroup="Platypus", resolve.root=T)

perms=RERconverge::getPermsContinuous(100, logAdultWeightcm, mamRERw, annotlist, toyTrees, mt)
corpermpvals=RERconverge::permpvalcor(res, perms)
enrichpermpvals=RERconverge::permpvalenrich(enrichment, perms)

# add permulations to real results

res$permpval=corpermpvals[match(rownames(res), names(corpermpvals))]
res$permpvaladj=p.adjust(res$permpval, method="BH")
count=1
while(count<=length(enrichment)){
  enrichment[[count]]$permpval=enrichpermpvals[[count]][match(rownames(enrichment[[count]]),
                                                              names(enrichpermpvals[[count]]))]
  enrichment[[count]]$permpvaladj=p.adjust(enrichment[[count]]$permpval, method="BH")
  count=count+1
}
```

As an alternative to running correlation and enrichment analyses simultaneously, the `getPermsContinuous` function may be used to run correlation analyses alone, and then the `getEnrichPermsContinuous` function may be used to calculate null enrichment statistics based on the null correlation statistics.  In this case, the output from `getEnrichPermsContinuous` should be supplied to `permpvalcor` and `permpvalenrich` functions.

This pipeline may be useful in the following cases
1. If very large datasets with many datasets are batched to run `readTrees`, `getAllResiduals`, and to calculate gene correlation statistics and then combined during standard RERconverge analysis, the following procedure should be used to mimic those steps during permulation analyses
2. If new pathway annotations are being tested after gene correlation permulations have already been run, `getEnrichPermsContinuous` can be used to calculate permulation enrichment statistics without rerunning correlation analyses.


```{r, message=F, warning=F, error=F, results=F}
permsnoenrich=RERconverge::getPermsContinuous(100, logAdultWeightcm, mamRERw, annotlist, toyTrees, mt, calculateenrich = F)
permswithenrich=RERconverge::getEnrichPermsContinuous(permsnoenrich, enrichment, annotlist)
corpermpvals2=RERconverge::permpvalcor(res, permswithenrich)
enrichpermpvals2=RERconverge::permpvalenrich(enrichment, permswithenrich)

plot(corpermpvals,corpermpvals2)
plot(enrichpermpvals[[1]],enrichpermpvals2[[1]])
```

Note variations in these two separate sets of permulation p-values.  This stochasticity highlights the necessity of running many permulations to explore as much null phenotype space as possible.

Permulations can also be run in batches and combined using `combinePermData`.  This function would be useful for combining several permulation batches run in tandem for computational efficiency.

```{r, message=F, warning=F, error=F, results=F}
perm2=RERconverge::getPermsContinuous(100, logAdultWeightcm, mamRERw, annotlist, toyTrees, mt)
combperms=RERconverge::combinePermData(perms, perm2)
```

Subsequent analyses should include subsetting observed significant pathways and genes according to parametric statistics based on significance according to permulation statistics.

